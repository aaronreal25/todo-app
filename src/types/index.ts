export type TaskStatus = 'PENDING' | 'IN_PROGRESS' | 'COMPLETED';

export interface Task {
  id: string;
  title: string;
  description: string;
  status: TaskStatus;
  dueDate: string;
  createdAt: string;
}

export interface TaskFilter {
  status: TaskStatus | 'ALL';
  searchQuery: string;
}

export interface TaskState {
  tasks: Task[];
  filter: TaskFilter;
  isLoading: boolean;
  error: string | null;
}

export type TaskAction =
  | { type: 'SET_TASKS'; payload: Task[] }
  | { type: 'ADD_TASK'; payload: Task }
  | { type: 'UPDATE_TASK'; payload: Task }
  | { type: 'DELETE_TASK'; payload: string }
  | { type: 'SET_FILTER'; payload: Partial<TaskFilter> }
  | { type: 'SET_LOADING'; payload: boolean }
  | { type: 'SET_ERROR'; payload: string | null };