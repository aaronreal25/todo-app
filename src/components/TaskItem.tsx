'use client';

import React from 'react';
import { format } from 'date-fns';
import { Task, TaskStatus } from '@/types';

interface TaskItemProps {
  task: Task;
  onStatusChange: (id: string, status: TaskStatus) => void;
  onEdit: (task: Task) => void;
  onDelete: (id: string) => void;
}

const TaskItem: React.FC<TaskItemProps> = React.memo(({ 
  task, 
  onStatusChange, 
  onEdit, 
  onDelete 
}) => {
  const currentDate = new Date();
  currentDate.setHours(0, 0, 0, 0);
  
  const dueDate = new Date(task.dueDate);
  dueDate.setHours(0, 0, 0, 0);

  const isOverdue = dueDate < currentDate;

  const getBorderColor = (status: TaskStatus): string => {
    switch (status) {
      case 'PENDING':
        return 'border-yellow-500';
      case 'IN_PROGRESS':
        return 'border-blue-500';
      case 'COMPLETED':
        return 'border-green-500';
      default:
        return 'border-gray-700';
    }
  };

  const getStatusColor = (status: TaskStatus): string => {
    switch (status) {
      case 'PENDING':
        return 'bg-yellow-900 text-yellow-100';
      case 'IN_PROGRESS':
        return 'bg-blue-900 text-blue-100';
      case 'COMPLETED':
        return 'bg-green-900 text-green-100';
      default:
        return 'bg-gray-800 text-gray-100';
    }
  };

  const borderColor = getBorderColor(task.status);
  const finalBorderClass = isOverdue && task.status !== 'COMPLETED' 
    ? 'border-red-500' 
    : borderColor;

  return (
    <div data-testid="task-item" className={`task-item p-4 mb-4 border-2 rounded-lg bg-gray-800 ${finalBorderClass}`}>
      <div className="task-item-header flex justify-between items-center mb-2">
        <h3 className="task-item-title text-lg font-semibold text-white">{task.title}</h3>
        <span 
          role="status"
          aria-label={task.status.toLowerCase()}
          className={`task-item-status px-2 py-1 rounded ${getStatusColor(task.status)}`}
        >
          {task.status.replace('_', ' ')}
        </span>
      </div>
      
      <p className="task-item-description text-gray-300 mb-2">{task.description}</p>
      <p className={`task-item-due-date text-sm ${isOverdue && task.status !== 'COMPLETED' ? 'text-red-400 font-semibold' : 'text-gray-400'}`}>
        Due: {format(new Date(task.dueDate), 'MMM dd, yyyy')}
      </p>

      <div className="task-item-controls flex justify-between mt-4">
        <select
          className="task-status-select p-2 border rounded bg-gray-700 text-white border-gray-600 focus:border-blue-500 focus:ring-blue-500"
          value={task.status}
          onChange={(e) => onStatusChange(task.id, e.target.value as TaskStatus)}
        >
          <option value="PENDING">Pending</option>
          <option value="IN_PROGRESS">In Progress</option>
          <option value="COMPLETED">Completed</option>
        </select>

        <div className="task-item-buttons space-x-2">
          <button
            onClick={() => onEdit(task)}
            className="task-edit-button px-3 py-1 bg-blue-600 text-white rounded hover:bg-blue-700"
          >
            Edit
          </button>
          <button
            onClick={() => onDelete(task.id)}
            className="task-delete-button px-3 py-1 bg-red-600 text-white rounded hover:bg-red-600"
          >
            Delete
          </button>
        </div>
      </div>
    </div>
  );
});

TaskItem.displayName = 'TaskItem';

export default TaskItem;
