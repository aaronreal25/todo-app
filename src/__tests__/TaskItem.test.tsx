import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import TaskItem from '@/components/TaskItem';
import { Task, TaskStatus } from '@/types';

const mockTask: Task = {
  id: '1',
  title: 'Test Task',
  description: 'Test Description',
  status: 'PENDING' as TaskStatus,
  dueDate: new Date().toISOString().split('T')[0],
  createdAt: new Date().toISOString(),
};

describe('TaskItem', () => {
  const mockOnStatusChange = jest.fn();
  const mockOnEdit = jest.fn();
  const mockOnDelete = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  // Interaction tests
  it('renders task details correctly', () => {
    render(
      <TaskItem
        task={mockTask}
        onStatusChange={mockOnStatusChange}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
      />
    );

    expect(screen.getByText('Test Task')).toBeInTheDocument();
    expect(screen.getByText('Test Description')).toBeInTheDocument();
    expect(screen.getByText('PENDING')).toBeInTheDocument();
  });

  it('calls onStatusChange when status is changed', () => {
    render(
      <TaskItem
        task={mockTask}
        onStatusChange={mockOnStatusChange}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
      />
    );

    fireEvent.change(screen.getByRole('combobox'), {
      target: { value: 'IN_PROGRESS' }
    });

    expect(mockOnStatusChange).toHaveBeenCalledWith('1', 'IN_PROGRESS');
  });

  it('calls handlers when buttons are clicked', () => {
    render(
      <TaskItem
        task={mockTask}
        onStatusChange={mockOnStatusChange}
        onEdit={mockOnEdit}
        onDelete={mockOnDelete}
      />
    );

    fireEvent.click(screen.getByRole('button', { name: /edit/i }));
    expect(mockOnEdit).toHaveBeenCalledWith(mockTask);

    fireEvent.click(screen.getByRole('button', { name: /delete/i }));
    expect(mockOnDelete).toHaveBeenCalledWith('1');
  });

  // Snapshot tests
  describe('Snapshots', () => {
    it('matches snapshot for pending task', () => {
      const { container } = render(
        <TaskItem
          task={mockTask}
          onStatusChange={mockOnStatusChange}
          onEdit={mockOnEdit}
          onDelete={mockOnDelete}
        />
      );
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot for in-progress task', () => {
      const inProgressTask = { ...mockTask, status: 'IN_PROGRESS' as TaskStatus };
      const { container } = render(
        <TaskItem
          task={inProgressTask}
          onStatusChange={mockOnStatusChange}
          onEdit={mockOnEdit}
          onDelete={mockOnDelete}
        />
      );
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot for completed task', () => {
      const completedTask = { ...mockTask, status: 'COMPLETED' as TaskStatus };
      const { container } = render(
        <TaskItem
          task={completedTask}
          onStatusChange={mockOnStatusChange}
          onEdit={mockOnEdit}
          onDelete={mockOnDelete}
        />
      );
      expect(container).toMatchSnapshot();
    });

    it('matches snapshot for overdue task', () => {
      const pastDate = new Date();
      pastDate.setDate(pastDate.getDate() - 1);
      const overdueTask = {
        ...mockTask,
        dueDate: pastDate.toISOString().split('T')[0]
      };
      const { container } = render(
        <TaskItem
          task={overdueTask}
          onStatusChange={mockOnStatusChange}
          onEdit={mockOnEdit}
          onDelete={mockOnDelete}
        />
      );
      expect(container).toMatchSnapshot();
    });
  });
});