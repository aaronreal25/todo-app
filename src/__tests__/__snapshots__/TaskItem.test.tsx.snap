// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskItem Snapshots matches snapshot for completed task 1`] = `
<div>
  <div
    class="task-item p-4 mb-4 border-2 rounded-lg bg-gray-800 border-green-500"
    data-testid="task-item"
  >
    <div
      class="task-item-header flex justify-between items-center mb-2"
    >
      <h3
        class="task-item-title text-lg font-semibold text-white"
      >
        Test Task
      </h3>
      <span
        aria-label="completed"
        class="task-item-status px-2 py-1 rounded bg-green-900 text-green-100"
        role="status"
      >
        COMPLETED
      </span>
    </div>
    <p
      class="task-item-description text-gray-300 mb-2"
    >
      Test Description
    </p>
    <p
      class="task-item-due-date text-sm text-gray-400"
    >
      Due: 
      Nov 28, 2024
    </p>
    <div
      class="task-item-controls flex justify-between mt-4"
    >
      <select
        class="task-status-select p-2 border rounded bg-gray-700 text-white border-gray-600 focus:border-blue-500 focus:ring-blue-500"
      >
        <option
          value="PENDING"
        >
          Pending
        </option>
        <option
          value="IN_PROGRESS"
        >
          In Progress
        </option>
        <option
          value="COMPLETED"
        >
          Completed
        </option>
      </select>
      <div
        class="task-item-buttons space-x-2"
      >
        <button
          class="task-edit-button px-3 py-1 bg-blue-600 text-white rounded hover:bg-blue-700"
        >
          Edit
        </button>
        <button
          class="task-delete-button px-3 py-1 bg-red-600 text-white rounded hover:bg-red-600"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`TaskItem Snapshots matches snapshot for in-progress task 1`] = `
<div>
  <div
    class="task-item p-4 mb-4 border-2 rounded-lg bg-gray-800 border-blue-500"
    data-testid="task-item"
  >
    <div
      class="task-item-header flex justify-between items-center mb-2"
    >
      <h3
        class="task-item-title text-lg font-semibold text-white"
      >
        Test Task
      </h3>
      <span
        aria-label="in_progress"
        class="task-item-status px-2 py-1 rounded bg-blue-900 text-blue-100"
        role="status"
      >
        IN PROGRESS
      </span>
    </div>
    <p
      class="task-item-description text-gray-300 mb-2"
    >
      Test Description
    </p>
    <p
      class="task-item-due-date text-sm text-gray-400"
    >
      Due: 
      Nov 28, 2024
    </p>
    <div
      class="task-item-controls flex justify-between mt-4"
    >
      <select
        class="task-status-select p-2 border rounded bg-gray-700 text-white border-gray-600 focus:border-blue-500 focus:ring-blue-500"
      >
        <option
          value="PENDING"
        >
          Pending
        </option>
        <option
          value="IN_PROGRESS"
        >
          In Progress
        </option>
        <option
          value="COMPLETED"
        >
          Completed
        </option>
      </select>
      <div
        class="task-item-buttons space-x-2"
      >
        <button
          class="task-edit-button px-3 py-1 bg-blue-600 text-white rounded hover:bg-blue-700"
        >
          Edit
        </button>
        <button
          class="task-delete-button px-3 py-1 bg-red-600 text-white rounded hover:bg-red-600"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`TaskItem Snapshots matches snapshot for overdue task 1`] = `
<div>
  <div
    class="task-item p-4 mb-4 border-2 rounded-lg bg-gray-800 border-red-500"
    data-testid="task-item"
  >
    <div
      class="task-item-header flex justify-between items-center mb-2"
    >
      <h3
        class="task-item-title text-lg font-semibold text-white"
      >
        Test Task
      </h3>
      <span
        aria-label="pending"
        class="task-item-status px-2 py-1 rounded bg-yellow-900 text-yellow-100"
        role="status"
      >
        PENDING
      </span>
    </div>
    <p
      class="task-item-description text-gray-300 mb-2"
    >
      Test Description
    </p>
    <p
      class="task-item-due-date text-sm text-red-400 font-semibold"
    >
      Due: 
      Nov 27, 2024
    </p>
    <div
      class="task-item-controls flex justify-between mt-4"
    >
      <select
        class="task-status-select p-2 border rounded bg-gray-700 text-white border-gray-600 focus:border-blue-500 focus:ring-blue-500"
      >
        <option
          value="PENDING"
        >
          Pending
        </option>
        <option
          value="IN_PROGRESS"
        >
          In Progress
        </option>
        <option
          value="COMPLETED"
        >
          Completed
        </option>
      </select>
      <div
        class="task-item-buttons space-x-2"
      >
        <button
          class="task-edit-button px-3 py-1 bg-blue-600 text-white rounded hover:bg-blue-700"
        >
          Edit
        </button>
        <button
          class="task-delete-button px-3 py-1 bg-red-600 text-white rounded hover:bg-red-600"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`TaskItem Snapshots matches snapshot for pending task 1`] = `
<div>
  <div
    class="task-item p-4 mb-4 border-2 rounded-lg bg-gray-800 border-yellow-500"
    data-testid="task-item"
  >
    <div
      class="task-item-header flex justify-between items-center mb-2"
    >
      <h3
        class="task-item-title text-lg font-semibold text-white"
      >
        Test Task
      </h3>
      <span
        aria-label="pending"
        class="task-item-status px-2 py-1 rounded bg-yellow-900 text-yellow-100"
        role="status"
      >
        PENDING
      </span>
    </div>
    <p
      class="task-item-description text-gray-300 mb-2"
    >
      Test Description
    </p>
    <p
      class="task-item-due-date text-sm text-gray-400"
    >
      Due: 
      Nov 28, 2024
    </p>
    <div
      class="task-item-controls flex justify-between mt-4"
    >
      <select
        class="task-status-select p-2 border rounded bg-gray-700 text-white border-gray-600 focus:border-blue-500 focus:ring-blue-500"
      >
        <option
          value="PENDING"
        >
          Pending
        </option>
        <option
          value="IN_PROGRESS"
        >
          In Progress
        </option>
        <option
          value="COMPLETED"
        >
          Completed
        </option>
      </select>
      <div
        class="task-item-buttons space-x-2"
      >
        <button
          class="task-edit-button px-3 py-1 bg-blue-600 text-white rounded hover:bg-blue-700"
        >
          Edit
        </button>
        <button
          class="task-delete-button px-3 py-1 bg-red-600 text-white rounded hover:bg-red-600"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;
